libqtile/window.py

```diff
@@ -1164,10 +1164,13 @@ def handle_ClientMessage(self, event):  # noqa: N802
                     logger.info("Focusing window")
                     self.qtile.current_screen.set_group(self.group)
                     self.group.focus(self)
-                elif focus_behavior == "smart" and self.group.screen and self.group.screen == self.qtile.current_screen:
-                    logger.info("Focusing window")
-                    self.qtile.current_screen.set_group(self.group)
-                    self.group.focus(self)
+                elif focus_behavior == "smart" and self.group.screen:
+                    if self.group.screen == self.qtile.current_screen:
+                        logger.info("Focusing window")
+                        self.qtile.current_screen.set_group(self.group)
+                        self.group.focus(self)
+                    else:
+                        logger.info("Ignoring focus request")
                 elif focus_behavior == "urgent" or (focus_behavior == "smart" and not self.group.screen):
```

Comment:
It looks like this is another case in this if-else chain that would break; can you fix this one up too?

This whole thing might be made cleaner by having a `needs_warning = True` that gets set to `False` in the various cases, vs. nesting things more. Whatever way you think looks best is fine with me.

Reply:
I am not quite sure if I understand you correctly? What would be messed up?
See #2242: I found a warning and traced it back to this piece of code. The only way I could figure out why this message was shown is the setup in the issue description (`focus_behavior == "smart" and self.group.screen and self.group.screen != self.qtile.current_screen` as this fell through before.
Could you clarify what might be broken and needs fixing?

Reply:
It looks like the next line in this if-else chane (line 1174, the one this comment thread is on) has the same bug as the one you fixed: it has additional tests in the if condition that would cause extra logging messages to be emitted when they shouldn't.

I think your fix is correct, and that there are additional instances of this same class of bug.

Reply:
Oh I see. The `focus_behavior == "smart"` branch is split into three parts which is quite unlucky: The third piece is on line 1174:
Assuming `focus_behavior == "smart"`, there are two cases:
Case A: `self.groups.screen == true`
  + checked: line 1167
  + Two sub-cases:  
    Case A.i: `self.groups.screen == self.qtile.current_screen` (line 1168)
    Case A.II: `self.groups.screen != self.qtile.current_screen` (line 1172)
Case B: `self.groups.screen == false`
  + checked: line 1174
So I don't think the setting "smart" should ever allow a fall-through, right?

Reply:
And "focus", "urgent" and "never" also seem to be fine <=> impossible to fall through

Reply:
Yes, exactly. It may be cleaner to have a `needs_warning` bool vs. rearranging all the if conditions, whatever looks best to you is fine with me.

Reply:
Hm. We want to throw a warning like [this](https://github.com/qtile/qtile/blob/9ffc1c6b8e540b56d9297ea3b19ec76fa0fcde0f/libqtile/window.py#L1177) one if `focus_behavior not in {"smart", "focus", "urgent", "never"}`.
Since we want to deal with all those cases to act appropriately and throw debugging messages, the `if`-`elif` block is necessary. I don't see how another local variable could be cleaner than just using the `else` case. The only thing I could imagine is to really group by `focus_behavior`, e. g. truly separate cases for "smart" etc. that would require one more level of nesting or a more complicated condition so that we would get away with the same level (also indentation level in this case) as in the pr at its current state.
So should I group the "smart" stuff together then?

Reply:
Sure, however it looks best to you sounds good.

Reply:
Okay, done in 8261115fad2335dfd7ddf794afbdf0321aae68db